cmd/controller/state/delta/controller.go

```diff
@@ -153,28 +154,29 @@ func (c *Controller) sendDeltas(ctx context.Context, firstDeltaReport bool) {
 	if firstDeltaReport {
 		meta = append(meta, "x-delta-full-snapshot", "true")
 	}
+
 	ctx = metadata.AppendToOutgoingContext(ctx, meta...)
 	deltaStream, err := c.castaiClient.KubernetesDeltaIngest(ctx, grpc.UseCompressor(gzip.Name))
 	if err != nil && !errors.Is(err, context.Canceled) {
-		c.log.Warnf("creating delta upload stream: %v", err)
-		return
+		return err
 	}
 	defer func() {
 		_ = deltaStream.CloseSend()
 	}()
 
 	var sentDeltasCount int
 	for _, item := range pendingDeltas {
+		item := item
 		pbItem := c.toCastaiDelta(item)
 		if err := c.sendDeltaItem(ctx, deltaStream, pbItem); err != nil {
-			c.log.Warnf("sending delta item: %v", err)
 			// Return any remaining items back to pending list.
-			c.upsertPendingItems(pendingDeltas[sentDeltasCount:])
-			return
+			c.upsertPendingItem(item)
```

Comment:
Any reason we no longer upsert the failing items? The only problem I could think of is, if there are events coming in the background. Here we should probably try to insert the item into the delta, only if it is missing.